src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -32,24 +33,119 @@ namespace
 {
     auto tid = celeritas::KernelParamCalculator::thread_id();
     if (!(tid < thread_range.size()))
         return;
     execute_thread(*(thread_range.cbegin() + tid.get()));
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the given executor using thread ids in the thread_range
+ */
+
+template<class F>
+__global__ void
+launch_action_impl(Range<ThreadId> const thread_range, F execute_thread)
+{
+    launch_kernel_impl(thread_range, execute_thread);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the given executor using thread ids in the thread_range with
+ \c __launch_bounds__
+ */
+
+#if CELERITAS_USE_CUDA
+template<class F, int T, int B = 1, int __B = (B * 32) / T>
+#elif CELERITAS_USE_HIP
+// see
+// https://rocm.docs.amd.com/projects/HIP/en/latest/reference/kernel_language.html#porting-from-cuda-launch-bounds
+template<class F, int T, int B = 1, int __B = B>
+#else
+#    error \
+        "Compiling device code without setting either CELERITAS_USE_CUDA or CELERITAS_USE_HIP"
+#endif
+__global__ void __launch_bounds__(T, __B)
+    launch_bounded_action_impl(Range<ThreadId> const thread_range,
+                               F execute_thread)
+{
+    launch_kernel_impl(thread_range, execute_thread);
+}
+
+// Select the correct kernel at compile time depending on the executor's launch
+// bounds.
+
+// instantiated if F doesn't define a member type F::Applier
+template<class F, std::enable_if_t<!has_applier_v<F>, bool> = true>
+constexpr auto select_kernel() -> decltype(&launch_action_impl<F>)
+{
+    constexpr auto ptr = &launch_action_impl<F>;
+    return ptr;
+}
+
+// instantiated if F::Applier has no launch bounds
+template<class F,
+         class __A = typename F::Applier,
+         std::enable_if_t<kernel_no_bound<__A>, bool> = true>
+constexpr auto select_kernel() -> decltype(&launch_action_impl<F>)
+{
+    constexpr auto ptr = &launch_action_impl<F>;
+    return ptr;
+}
+
+// instantiated if F::Applier defines one argument for launch bound
+template<class F,
+         class __A = typename F::Applier,
+         std::enable_if_t<kernel_max_blocks<__A>, bool> = true>
+constexpr auto select_kernel()
+    -> decltype(&launch_bounded_action_impl<F, __A::max_block_size>)
+{
+    constexpr auto ptr = &launch_bounded_action_impl<F, __A::max_block_size>;
+    return ptr;
+}
+
+// instantiated if F::Applier defines two arguments for launch bounds
+template<class F,
+         class __A = typename F::Applier,
+         std::enable_if_t<kernel_max_blocks_min_warps<__A>, bool> = true>
+constexpr auto select_kernel()
+    -> decltype(&launch_bounded_action_impl<F,
+                                            __A::max_block_size,
+                                            __A::min_warps_per_eu>)
+{
+    constexpr auto ptr
+        = &launch_bounded_action_impl<F, __A::max_block_size, __A::min_warps_per_eu>;
+    return ptr;
+}
+
```

Comment:
The preface code here is getting long and complicated enough to obscure the actual `ActionLauncher` class. Can you move all of this to the `detail/` directory and put it in a `namespace detail { `?